src/commitments/pedersen_BLS12_381.rs

```diff
@@ -0,0 +1,257 @@
+extern crate amcl;
+extern crate rand;
+
+use self::rand::rngs::EntropyRng;
+
+use self::amcl::bls381::ecp::ECP;
+use self::amcl::bls381::big::{BIG, MODBYTES};
+
+use hash_functions::HashFunction;
+use hash_functions::bls12_381_hash::BLS12_381_SHA256_G1;
+
+use commitments::CommitmentScheme;
+use commitments::CommitmentError;
+use utils::random::random_big_number;
+use amcl::bls381::rom::CURVE_ORDER;
+
+const SETUP_SEED_G1: &'static str = "Hyperledger-Cryptolib-Pedersen-Commitment-BLS-12-381-G1";
+const GROUP_G1_SIZE: usize = 2 * MODBYTES + 1;
+
+struct PedersenCommitmentBLS12_381_SHA256_G1 {}
+
+impl CommitmentScheme for PedersenCommitmentBLS12_381_SHA256_G1 {
+    // Returns `num_elements` + 1 generators. This is useful when committing to several messages say
+    // m_1, m_2, m_3, and so on. `setup` will this output g_1, g_2, g_3, g_4 and so on which can then be
+    // used for commitment f(g_1, g_2, g_3, g_4, ..., m_1, m_2, m_3...)
+    fn setup(num_elements: usize) -> Result<Vec<Vec<u8>>, CommitmentError> {
+        if num_elements == 0 {
+            return Err(CommitmentError::ZeroCountInPedersenSetup(String::from("num_elements cannot be 0")))
+        }
+        let mut generators: Vec<ECP> = vec![];
+        let mut hf = BLS12_381_SHA256_G1::new(None)?;
+        hf.update(SETUP_SEED_G1.as_bytes());
+        for _ in 0..num_elements+1 {
+            generators.push(hf.hash_on_group());
+            let m = hf.digest(None)?;
+            hf.update(&m);
+        }
+        Ok(generators.iter_mut().map(|g| {
+            let mut temp: [u8; GROUP_G1_SIZE] = [0; GROUP_G1_SIZE];
+            g.tobytes(&mut temp, false);
+            temp.to_vec()
+            }).collect()
+        )
+    }
+
+    fn commit(generators: &[&[u8]], messages: &[&[u8]]) -> Result<(Vec<u8>, Vec<u8>), CommitmentError> {
+        match Self::verify_generator_message_count(generators, messages) {
+            Ok(_) => (),
+            Err(err) => return Err(err)
+        }
+
+        let mut commitment = Self::commit_to_messages(generators, messages)?;
+        let idx = messages.len();
+        let gen = Self::get_point_from_bytes(generators[idx])?;
+        let mut blinding_factor_bytes: [u8; MODBYTES] = [0; MODBYTES];
+
+        let mut blinding_factor = random_big_number(&CURVE_ORDER, None);
+        commitment.add(&gen.mul(&blinding_factor));
+
+        blinding_factor.tobytes(&mut blinding_factor_bytes);
+
+        // Zero the bytes of blinding factor.
+        // TODO: A better mechanism is needed. Use either libsodium or some other library that makes sure that `blinding_factor`
+        // was never swapped to disk or was core dumped.
+        blinding_factor.zero();
+
```

Comment:
sodium_mlock I believe does this and is exposed with libsodium-ffi.
See https://download.libsodium.org/doc/helpers/memory_management.html

Reply:
I know, but i want to explore a more thorough mechanism (or library) and implement the mechanism at one place and use it everywhere. The ones on my list are [clear_on_drop](https://github.com/cesarb/clear_on_drop) and [secrets](https://github.com/stouset/secrets)